README.md
# Leaflet.TileLayer.GL

A [LeafletJS](http://www.leafletjs.com) plugin to appy WebGL shaders to your tiles.

With this plugin, you can apply colour transforms to your tiles, merge two or
more tiles with a custom function, perform on-the-fly hillshading, or create synthetic
tile layers based only on the map coordinates.

## Demo

#### On-the-fly colouring

This demo loads the ["toner" map style by Stamen](http://maps.stamen.com/toner/) and changes the colours on-the-fly:

[http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-antitoner.html](http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-antitoner.html)

#### Mandelbrot set

This demo loads no tiles and uses the map coordinates to draw a fractal set:

[http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-mandelbrot.html](http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-mandelbrot.html)


#### Flood & height

This demo uses [MapBox's "Terrain-RGB" tiles](https://www.mapbox.com/blog/terrain-rgb/) to play with the elevation: areas are coloured depending to the elevation (below 0 meters, between 0 and 5 meters, between 5 and 10 meters, above 10 meters).

[http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-flood.html](http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-flood.html)

#### Hypsometric tint

This demo uses [MapBox's "Terrain-RGB" tiles](https://www.mapbox.com/blog/terrain-rgb/) and applies a basic [hypsometric tint](https://en.wikipedia.org/wiki/Hypsometric_tints) colour ramp.

[http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-hypsometric.html](http://ivansanchez.gitlab.io/Leaflet.TileLayer.GL/demo/demo-hypsometric.html)


## Why?

Leaflet has been lagging behind when it comes to WebGL technology. Other map libraries (such as [OpenLayers 3]() and most notably [Tangram](https://mapzen.com/products/tangram/)) can already use WebGL shaders to apply transformations to map tiles and do fancy stuff.

The inflexion point are [MapBox's "Terrain-RGB" tiles](https://www.mapbox.com/blog/terrain-rgb/). WebGL manipulation of these tiles can provide real-time terrain relief and hill shading.

This takes some inspiration from [shadertoy.com](http://www.shadertoy.com), in the sense that the shaders work on two triangles with some predefined attributes and uniforms.

## Compatibility

Leaflet 1.0.1 (or newer), and a web browser that supports both [WebGL](http://caniuse.com/#search=webgl) and [ES6 `Promise`s](http://caniuse.com/#search=promise). You can also use a `Promise` polyfill for IE11.

## Usage

Include Leaflet and Leaflet.TileLayer.GL in your HTML:

```
<link rel="stylesheet" href="https://unpkg.com/leaflet@1.0.1/dist/leaflet.css" />
<script src="https://unpkg.com/leaflet@1.0.1/dist/leaflet.js"></script>
<script src='https://unpkg.com/leaflet.tilelayer.gl@latest/src/Leaflet.TileLayer.GL'></script>
```

Alternatively, fetch a local copy of Leaflet and Leaflet.TileLayer.GL with `npm install --save leaflet; npm install --save leaflet.tilelayer.gl` or `yarn add leaflet; yarn add leaflet.tilelayer.gl`

You can create instances of `L.TileLayer.GL` in your code. These take three new options: `vertexShader`, `fragmentShader` and `tileUrls`, e.g.:

```
	var antitoner = L.tileLayer.gl({
		vertexShader: "// String with GLSL vertex shader code",
		fragmentShader: "// String with GLSL fragment shader code",
		tileUrls: ['http://{s}.tile.stamen.com/toner/{z}/{x}/{y}.png']
	}).addTo(map);
```

Using this plugin requires some knowledge of WebGL and GLSL shaders. If you've never heard the terms "vertex shader" or "fragment shader", read [this WebGL tutorial](https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Tutorial/Getting_started_with_WebGL) to become acquinted, or [The Book Of Shaders](https://thebookofshaders.com/) to learn to do cool shaders, or [WebGL Fundamentals](http://webglfundamentals.org/webgl/lessons/webgl-image-processing.html) to see some WebGL image processing techniques.

The `vertexShader` and `fragmentShader` options contain shader code, in strings. For every map tile, the shaders will run once, on two triangles. This plugin does not allow you to create more triangles.

The vertex shader receives the following **attributes**:

* `aCRSCoords`: a `vec2` containing the *map* coordinates for the vertices (with values like `LatLng`s).
* `aVertexCoords`: a `vec2` containing the *viewport* coordinates for the vertices (with values from `-1.0` to `+1.0`). These make the two triangles span a whole tile. Make sure to use these as `gl_Position` in the vertex shader.
* `aTextureCoords`: a `vec2` containing the *texture* coordinates for the vertices. Use this for fetching texels.

The shaders receive the following **uniforms**:

* `uTexture0`: a `sampler2D` referring to the first loaded tile image. This exists only if the `tileUrls` option is not empty.
* `uTexture1`..`uTexture7`: texture samplers for the 2nd through 8th image.

## Demo shaders

This is the code used in the "antitoner" demo, commented and explained:

```js
// Create the vertex shader as a multi-line string. Note the "`" character, valid only in ES6 JavaScript.
// Shaders can be defined elsewhere, or loaded from other files or from the network,
// but they must be strings when used in a TileLayer.GL.
var antiTonerVertexShader = `
	// Define which attribs we want. In this case, aCRSCoords is not needed.
	attribute vec2 aVertexCoords;
	attribute vec2 aTextureCoords;

	// We need a varying so the fragment shader can use the interpolated aTextureCoords
	varying vec2 vTextureCoords;

	void main(void) {
		gl_Position = vec4(aVertexCoords , 1.0, 1.0);	// Use the vertex coords as given
		vTextureCoords = aTextureCoords;	// Pass the texture coords to the frag shader
	}
`

var antiTonerFragmentShader = `
	// The fragment shader runs once per fragment (AKA "once per screen pixel")
	precision highp float;
	uniform sampler2D uTexture0;	// This contains a reference to the tile image loaded from the network
	varying vec2 vTextureCoords;	// This is the interpolated texel coords for this fragment

	void main(void) {
		// Classic texel look-up (fetch the texture "pixel" color for this fragment)
		vec4 texelColour = texture2D(uTexture0, vec2(vTextureCoords.s, vTextureCoords.t));

		// If uncommented, this would output the image "as is"
		// gl_FragColor = texelColour;

		// Let's mix the colours a little bit, inverting the red and green channels.
		gl_FragColor = vec4(1.0 - texelColour.rg, texelColour.b, 1.0);
	}
`

// Instantiate our L.TileLayer.GL...
var antitoner = L.tileLayer.gl({
	// ... with the shaders we just wrote above...
	vertexShader: antiTonerVertexShader,
	fragmentShader: antiTonerFragmentShader,

	// ...and loading tile images from Stamen Toner as "uTexture0".
	// If this array contained more than one tile template string,
	// there would be "uTexture1", "uTexture2" and so on.
	tileUrls: ['http://{s}.tile.stamen.com/toner/{z}/{x}/{y}.png']
}).addTo(map);
```

## Cool things that should be doable, but nobody has yet shown interest in asking about, much less in implementing them

* Updating the shaders
* Reusing the same WebGL context for more than one `TileLayer.GL` (as the render
  calls are sync)
* Custom uniforms
* Time uniform(s)
* Some kind of animations (keep the loaded images in memory, implement a rendering
  loop)

## Legalese

----------------------------------------------------------------------------

"THE BEER-WARE LICENSE":
<ivan@sanchezortega.es> wrote this file. As long as you retain this notice you
can do whatever you want with this stuff. If we meet some day, and you think
this stuff is worth it, you can buy me a beer in return.

----------------------------------------------------------------------------



CHANGELOG

# 1.0.1

* Switch to subclassing from `GridLayer` instead of `TileLayer` to properly
  handle async draws and remove glitching.
* Prettified filenames and updated refs in `package.json`.
* Fixed CRS coords orientation *again* because GL coords systems are insane.
* Linting.

# 1.0.0

* Initial release
* Three demos (mandelbrot, terrain-rbg, antitoner)



src/Leaflet.GridLayer.GL.js
/*
 * @class GridLayer.GL
 * @inherits GridLayer
 *
 * This `GridLayer` runs some WebGL code on each grid cell, and puts an image
 * with the result back in place.
 * 
 * The contents of each cell can be purely synthetic (based only on the cell
 * coordinates), or be based on some remote tiles (used as textures in the WebGL
 * shaders).
 * 
 * The fragment shader is assumed to receive two `vec2` attributes, with the CRS
 * coordinates and the texture coordinates: `aCRSCoords` and `aTextureCoords`.
 * If textures are used, they are accesed through the uniforms `uTexture0` through `uTexture7`
 * There will always be four vertices forming two triangles (a quad).
 * 
 */


L.TileLayer.GL = L.TileLayer.extend({

	options: {
		// @option tileUrls: Array
		// Array of tile URL templates (as in `L.TileLayer`), between zero and 8 elements.
		tileUrls: [],
		
		// @option vertexShader: String
		// A string representing the GLSL vertex shader to be run.
		vertexShader: '',
		
		// @option fragmentShader: String
		// A string representing the GLSL fragment shader to be run.
		fragmentShader: ''
	},
	
	// On instantiating the layer, it will initialize all the GL context
	//   and upload the shaders to the GPU, along with the vertex buffer
	//   (the vertices will stay the same for all tiles).
	initialize: function(options) {
		options = L.setOptions(this, options);
		
		this._renderer = L.DomUtil.create('canvas');
		this._renderer.width = this._renderer.height = this.options.tileSize;
		
		var gl = this._gl = this._renderer.getContext('webgl', {
			premultipliedAlpha: false
		}) || this._renderer.getContext("experimental-webgl", {
			premultipliedAlpha: false
		});
		gl.viewportWidth  = this.options.tileSize;
		gl.viewportHeight = this.options.tileSize;
		
		this._loadGLProgram();
		
		// Init textures
		this._textures = [];
		for (var i=0; i<this.options.tileUrls.length && i<8; i++) {
			gl.bindTexture(gl.TEXTURE_2D, this._textures[i] = gl.createTexture());
			gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, this._size, this._size, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
			
			gl.uniform1i(gl.getUniformLocation(this._glProgram, "uTexture" + i), i);
		}
		
	},

	
	_loadGLProgram: function() {
		var gl = this._gl;
		
		var program = this._glProgram = gl.createProgram();
		var vertexShader   = gl.createShader(gl.VERTEX_SHADER);
		var fragmentShader = gl.createShader(gl.FRAGMENT_SHADER);
		gl.shaderSource(vertexShader, this.options.vertexShader);
		gl.shaderSource(fragmentShader, this.options.fragmentShader);
		gl.compileShader(vertexShader);
		gl.compileShader(fragmentShader);
		if (!gl.getShaderParameter(vertexShader, gl.COMPILE_STATUS)) {
			console.error(gl.getShaderInfoLog(vertexShader));
			return null;
		}
		if (!gl.getShaderParameter(fragmentShader, gl.COMPILE_STATUS)) {
			console.error(gl.getShaderInfoLog(fragmentShader));
			return null;
		}
		gl.attachShader(program, vertexShader);
		gl.attachShader(program, fragmentShader);
		gl.linkProgram(program);
		gl.useProgram(program);
		
		// There will be two vec2 vertex attributes per vertex - aCRSCoords and
		// aTextureCoords
		this._aVertexPosition = gl.getAttribLocation(program, "aVertexCoords");
		this._aTexPosition = gl.getAttribLocation(program, "aTextureCoords");
		this._aCRSPosition = gl.getAttribLocation(program, "aCRSCoords");
		
// 		console.log('Tex position: ', this._aTexPosition);
// 		console.log('CRS position: ', this._aCRSPosition);
		
		// Create a data buffer with 16 elements - the (easting,northing) CRS coords
		// and the (s,t) texture coords for each of the 4 vertices
		this._CRSBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._CRSBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(8), gl.STATIC_DRAW);
		if (this._aCRSPosition !== -1) {
			gl.enableVertexAttribArray(this._aCRSPosition);
			gl.vertexAttribPointer(this._aCRSPosition, 2, gl.FLOAT, false, 8, 0);
		}
		
		this._TexCoordsBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._TexCoordsBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([
			1.0, 0.0,
			0.0, 0.0,
			1.0, 1.0,
			0.0, 1.0,
		]), gl.STATIC_DRAW);
		if (this._aTexPosition !== -1) {
			gl.enableVertexAttribArray(this._aTexPosition);
			gl.vertexAttribPointer(this._aTexPosition, 2, gl.FLOAT, false, 8, 0);
		}
		
		this._VertexCoordsBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._VertexCoordsBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([
			 1,  1,
			-1,  1,
			 1, -1,
			-1, -1
		]), gl.STATIC_DRAW);
		if (this._aVertexPosition !== -1) {
			gl.enableVertexAttribArray(this._aVertexPosition);
			gl.vertexAttribPointer(this._aVertexPosition, 2, gl.FLOAT, false, 8, 0);
		}
		
	},
	
	// This is called once per tile - uses the layer's GL context to
	//   render a tile, passing the complex space coordinates to the
	//   GPU, and asking to render the vertexes (as triangles) again.
	// Every pixel will be opaque, so there is no need to clear the scene.
	_render: function(coords, imgs) {
		var gl = this._gl;
		gl.viewport(0, 0, gl.drawingBufferWidth, gl.drawingBufferHeight);
		gl.clearColor(0.5, 0.5, 0.5, 0);
		gl.enable(gl.BLEND);
		
		var tileBounds = this._tileCoordsToBounds(coords);
		
		// Create data array for CRS buffer
		var data = [
			// Vertex 0
			tileBounds.getEast(), tileBounds.getSouth(),

			// Vertex 1
			tileBounds.getWest(), tileBounds.getSouth(),

			// Vertex 2
			tileBounds.getEast(), tileBounds.getNorth(),

			// Vertex 3
			tileBounds.getWest(), tileBounds.getNorth(),
		];
		
		// ...upload them to the GPU...
		gl.bindBuffer(gl.ARRAY_BUFFER, this._CRSBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(data), gl.STATIC_DRAW);
		
		// ... and then the magic happens.
		gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
		
	},
	
	
	createTile: function (coords, done) {
		var tile = L.DomUtil.create('canvas', 'leaflet-tile');
		tile.width = tile.height = this.options.tileSize;
		tile.onselectstart = tile.onmousemove = L.Util.falseFn;
		
		var ctx = tile.getContext('2d');
		if (this.options.tileUrls.length === 0) {
			this._render(coords);
			ctx.drawImage(this._renderer, 0, 0);
			setTimeout(done, 50);
		} else {
			var texFetches = [];
			for (var i=0; i<this.options.tileUrls.length && i<8; i++) {
				this._url = this.options.tileUrls[i];
				
				texFetches.push(new Promise(function(resolve,reject){
					var tile = document.createElement('img');
					tile.crossOrigin = '';
					tile.src = this.getTileUrl(coords);
					L.DomEvent.on(tile, 'load', resolve.bind(this, tile));
					L.DomEvent.on(tile, 'error', reject.bind(this, tile));
				}.bind(this)));
			}
			
			Promise.all(texFetches)
			.catch(this._tileOnError)
			.then(function(textureImages){
				console.log(textureImages);
				var gl = this._gl;
				for (var i=0; i<this.options.tileUrls.length && i<8; i++) {
					gl.bindTexture(gl.TEXTURE_2D, this._textures[i]);
					gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, textureImages[i]);
					gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
					gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_NEAREST);
					gl.generateMipmap(gl.TEXTURE_2D);
				}
				
				this._render(coords);
				ctx.drawImage(this._renderer, 0, 0);
				done();
			}.bind(this));
		}
	
		return tile;
	},
	
	
});


L.tileLayer.gl = function(opts) {
	return new L.TileLayer.GL(opts);
}


src/Leaflet.TileLayer.GL.js
/*
 * @class GridLayer.GL
 * @inherits GridLayer
 *
 * This `GridLayer` runs some WebGL code on each grid cell, and puts an image
 * with the result back in place.
 *
 * The contents of each cell can be purely synthetic (based only on the cell
 * coordinates), or be based on some remote tiles (used as textures in the WebGL
 * shaders).
 *
 * The fragment shader is assumed to receive two `vec2` attributes, with the CRS
 * coordinates and the texture coordinates: `aCRSCoords` and `aTextureCoords`.
 * If textures are used, they are accesed through the uniforms `uTexture0` through `uTexture7`
 * There will always be four vertices forming two triangles (a quad).
 *
 */


L.TileLayer.GL = L.GridLayer.extend({

	options: {
		// @option tileUrls: Array
		// Array of tile URL templates (as in `L.TileLayer`), between zero and 8 elements.
		tileUrls: [],

		// @option vertexShader: String
		// A string representing the GLSL vertex shader to be run.
		vertexShader: '',

		// @option fragmentShader: String
		// A string representing the GLSL fragment shader to be run.
		fragmentShader: '',

		subdomains: ['a', 'b', 'c', 'd']
	},

	// On instantiating the layer, it will initialize all the GL context
	//   and upload the shaders to the GPU, along with the vertex buffer
	//   (the vertices will stay the same for all tiles).
	initialize: function (options) {
		options = L.setOptions(this, options);

		this._renderer = L.DomUtil.create('canvas');
		this._renderer.width = this._renderer.height = options.tileSize;

		var gl = this._gl = this._renderer.getContext('webgl', {
			premultipliedAlpha: false
		}) || this._renderer.getContext("experimental-webgl", {
			premultipliedAlpha: false
		});
		gl.viewportWidth  = options.tileSize;
		gl.viewportHeight = options.tileSize;

		this._loadGLProgram();

		// Init textures
		this._textures = [];
		for (var i=0; i<options.tileUrls.length && i<8; i++) {
			gl.activeTexture(gl.TEXTURE0 + i);
			gl.bindTexture(gl.TEXTURE_2D, this._textures[i] = gl.createTexture());
			gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, this._size, this._size, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
			gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);

			gl.uniform1i(gl.getUniformLocation(this._glProgram, "uTexture" + i), i);
		}

	},


	_loadGLProgram: function () {
		var gl = this._gl;

		var program = this._glProgram = gl.createProgram();
		var vertexShader   = gl.createShader(gl.VERTEX_SHADER);
		var fragmentShader = gl.createShader(gl.FRAGMENT_SHADER);
		gl.shaderSource(vertexShader, this.options.vertexShader);
		gl.shaderSource(fragmentShader, this.options.fragmentShader);
		gl.compileShader(vertexShader);
		gl.compileShader(fragmentShader);
		if (!gl.getShaderParameter(vertexShader, gl.COMPILE_STATUS)) {
			console.error(gl.getShaderInfoLog(vertexShader));
			return null;
		}
		if (!gl.getShaderParameter(fragmentShader, gl.COMPILE_STATUS)) {
			console.error(gl.getShaderInfoLog(fragmentShader));
			return null;
		}
		gl.attachShader(program, vertexShader);
		gl.attachShader(program, fragmentShader);
		gl.linkProgram(program);
		gl.useProgram(program);

		// There will be two vec2 vertex attributes per vertex - aCRSCoords and
		// aTextureCoords
		this._aVertexPosition = gl.getAttribLocation(program, "aVertexCoords");
		this._aTexPosition = gl.getAttribLocation(program, "aTextureCoords");
		this._aCRSPosition = gl.getAttribLocation(program, "aCRSCoords");

// 		console.log('Tex position: ', this._aTexPosition);
// 		console.log('CRS position: ', this._aCRSPosition);

		// Create three data buffer with 8 elements each - the (easting,northing)
		// CRS coords, the (s,t) texture coords and the viewport coords for each
		// of the 4 vertices
		// Data for the texel and viewport coords is totally static, and
		// needs to be declared only once.
		this._CRSBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._CRSBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(8), gl.STATIC_DRAW);
		if (this._aCRSPosition !== -1) {
			gl.enableVertexAttribArray(this._aCRSPosition);
			gl.vertexAttribPointer(this._aCRSPosition, 2, gl.FLOAT, false, 8, 0);
		}

		this._TexCoordsBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._TexCoordsBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([
			1.0, 0.0,
			0.0, 0.0,
			1.0, 1.0,
			0.0, 1.0,
		]), gl.STATIC_DRAW);
		if (this._aTexPosition !== -1) {
			gl.enableVertexAttribArray(this._aTexPosition);
			gl.vertexAttribPointer(this._aTexPosition, 2, gl.FLOAT, false, 8, 0);
		}

		this._VertexCoordsBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this._VertexCoordsBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([
			 1,  1,
			-1,  1,
			 1, -1,
			-1, -1
		]), gl.STATIC_DRAW);
		if (this._aVertexPosition !== -1) {
			gl.enableVertexAttribArray(this._aVertexPosition);
			gl.vertexAttribPointer(this._aVertexPosition, 2, gl.FLOAT, false, 8, 0);
		}

	},

	// This is called once per tile - uses the layer's GL context to
	//   render a tile, passing the complex space coordinates to the
	//   GPU, and asking to render the vertexes (as triangles) again.
	// Every pixel will be opaque, so there is no need to clear the scene.
	_render: function (coords) {
		var gl = this._gl;
		gl.viewport(0, 0, gl.drawingBufferWidth, gl.drawingBufferHeight);
		gl.clearColor(0.5, 0.5, 0.5, 0);
		gl.enable(gl.BLEND);

		var tileBounds = this._tileCoordsToBounds(coords);

		// Create data array for CRS buffer
		var data = [
			// Vertex 0
			tileBounds.getEast(), tileBounds.getNorth(),

			// Vertex 1
			tileBounds.getWest(), tileBounds.getNorth(),

			// Vertex 2
			tileBounds.getEast(), tileBounds.getSouth(),

			// Vertex 3
			tileBounds.getWest(), tileBounds.getSouth(),
		];

		// ...upload them to the GPU...
		gl.bindBuffer(gl.ARRAY_BUFFER, this._CRSBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(data), gl.STATIC_DRAW);

		// ... and then the magic happens.
		gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);

	},

	// Monkey-patch some of the L.TileLayer methods so we can call getTileUrl
	// without errors
	_getSubdomain: L.TileLayer.prototype._getSubdomain,
	_getZoomForUrl: function () { return this._tileZoom; },

	createTile: function (coords, done) {
		var tile = L.DomUtil.create('canvas', 'leaflet-tile');
		tile.width = tile.height = this.options.tileSize;
		tile.onselectstart = tile.onmousemove = L.Util.falseFn;

		var ctx = tile.getContext('2d');
		if (this.options.tileUrls.length === 0) {
			this._render(coords);
			ctx.drawImage(this._renderer, 0, 0);
			setTimeout(done, 50);
		} else {
			var texFetches = [];
			for (var i=0; i<this.options.tileUrls.length && i<8; i++) {
				this._url = this.options.tileUrls[i];

				texFetches.push(new Promise(function (resolve,reject) {
					var tile = document.createElement('img');
					tile.crossOrigin = '';
					tile.src = L.TileLayer.prototype.getTileUrl.call(this, coords);
					L.DomEvent.on(tile, 'load', resolve.bind(this, tile));
					L.DomEvent.on(tile, 'error', reject.bind(this, tile));
				}.bind(this)));
			}

			Promise.all(texFetches)
			.catch(this._tileOnError)
			.then(function (textureImages) {
// 				console.log(textureImages);
				var gl = this._gl;
				for (var i=0; i<this.options.tileUrls.length && i<8; i++) {
					gl.activeTexture(gl.TEXTURE0 + i);
					gl.bindTexture(gl.TEXTURE_2D, this._textures[i]);
					gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, textureImages[i]);
					gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
					gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_NEAREST);
					gl.generateMipmap(gl.TEXTURE_2D);
				}

				this._render(coords);
				ctx.drawImage(this._renderer, 0, 0);
				done();
			}.bind(this));
		}

		return tile;
	},


});


L.tileLayer.gl = function (opts) {
	return new L.TileLayer.GL(opts);
};
